### ISSUE TITLE
Unable to conditionally render Table Columns and Cells in `react-aria-components@1.8.0`

### ISSUE THREAD

**Issue by jaknas** [adobe/react-spectrum#8127]
### Provide a general summary of the issue here

After updating to `react-aria-components@1.8.0`, we noticed that the Table component throws `Cell count must match column count. Found 5 cells and 3 columns.` when trying to remove some columns and cells conditionally. One use-case of that could be reducing the amount of columns and cells based on viewport size.

### 🤔 Expected Behavior?

Columns and cells can be added/removed conditionally.

### 😯 Current Behavior

Error message saying `Cell count must match column count. Found 5 cells and 3 columns.` is rendered. Amount of cells and columns is variable depending on how much of those we want to hide/show.

### 🔦 Context

Is there another way we should be approaching this pattern, instead of modifying the amount of columns and cells being rendered? The use-case we have right now is hiding the amount of columns in mobile view. Has this use case come up in react-spectrum/s2? How did you account for this in code/styles?

### 🖥️ Steps to Reproduce

| `react-aria-components@1.7.1` | `react-aria-components@1.8.0` |
|--------|--------|
|https://codesandbox.io/p/sandbox/ractable-1-7-1-forked-h29l3v | https://codesandbox.io/p/sandbox/ractable-1-8-0-vv6qld | 

After the sandbox is opened, click the "Hide Columns" button. 

In 1.7.1 notice that the table re-renders just fine

In 1.8.0 I receive a runtime error.

What I think is important to note, is when in `1.7.1` I render my rows via `TableBody` render prop, I also receive this error. 

### Version

1.8.0

### What browsers are you seeing the problem on?

Chrome

### What operating system are you using?

MacOS

**Comment by wilson-nickolas** [comment]
@jaknas 
I came across your issue and thought I'd take a look.
After looking at your examples, I noticed a few things.
- You should be using the correct way for dynamic collections
  - https://react-spectrum.adobe.com/react-aria/Table.html#content
- With your primary use case being to hide columns on view point size, I'd advise you don't remove the column from the collection. Instead use styling to hide it.

Anyway regarding the main issue, I too initially thought there was a problem with `row.columns` not passing the updated column list to row items. However, I took a look at the source  and found there is a prop called `dependencies` which seems to be designed to invalidate the cached row and cells. It took a bit of time to work out (no docs about it - what a surprise) but if you pass the value that changes just like you would for a hook dependency, it will cause the row to rerender with the updated columns. 

Here is an example that I forked and modified from your original. Hope this helps!

https://codesandbox.io/p/devbox/ractable-1-8-0-v2-forked-forked-tgy4pk?workspaceId=ws_K2rfy5Y7dqbzg9UnP2265z

**Comment by lukpsaxo** [comment]
We are getting this too. I've spent a day debugging and learning about Document/Collections.
We already pass dependencies down.. I can't figure out yet why I cannot create a reproducible sandbox example but I can within our app :/

**Comment by snowystinger** [comment]
Thanks for investigating so far, very helpful in jump-starting this work. I'm just letting people know we are looking into this, fingers crossed we'll find a fix before release.

Unfortunately, it's not quite as simple as just the change outlined in `removeChild`, though that change is likely still needed.
In Tables, columns and items are separate, and now we need to mark the rows/items as dirty/changed/stale based on the columns, which isn't completely trivial. So we're looking into ways to link these two structures or communicate that the rows need a re-render. Previously, without the caching we implemented to speed up large collections, we were getting lucky that the entire collection was updating for every change, no matter how trivial or far-reaching.

**Comment by snowystinger** [comment]
linking related ticket [Updating columns in S2 TableView throws an error](https://github.com/adobe/react-spectrum/issues/7345#top)
